### DISCUSSION TITLE
Changing the location of download (cache)

### DISCUSSION THREAD

**Discussion by Xelbayria** [ScoopInstaller/Scoop#5498] [Help and Questions]
Hello, 

is it possible to change the location `~/scoop/cache` to say `C:\Users\Username\Downloads`? 

something like this in the json: 
`"download_to": "$Users$\Downloads"`

So scoop would download the file to `Downloads folder` instead of `cache folder`

**Comment by HUMORCE** [comment]
```
❯ scoop help config
...
cache_path:
      For downloads, defaults to 'cache' folder under Scoop root directory.
...
```

**Reply by Xelbayria** [reply]
I would like to use the default download for other software... but there are a few JSON files that I would like to specific to the different location for downloading? 

is there a way to make this possible?

**Reply by HUMORCE** [reply]
This breaks management of downloaded content.

> is there a way to make this possible?

You can download file to any location with scripts then handle them with function(e.g. extraction) of Scoop core, but I think it's a stupid approach.

**Reply by Xelbayria** [reply]
> This breaks management of downloaded content.
> 
> > is there a way to make this possible?
> 
> You can download file to any location with scripts then handle them with function(e.g. extraction) of Scoop core, but I think it's a stupid approach.

you may think it's a stupid approach. but i have my reasons for that. I don't blame you for that. so scripts, you meant `"script"` to move the downloaded file to a different location and that's it??? 
The thing is that I only want to use `scoop download`, that's all.

**Reply by Xelbayria** [reply]
okay.. it seems the simple solution is just to make a copy of `scoop-download.ps1` and modify it to `scoop-get.ps1` or something. and change the cache_path to different location. 

is there a way to exclude `scoop-get.ps1` so i won't get the request to push the folder of scoop to github? and the file won't be affected or removed if scoop happened to update to the latest version?

is it possible to do that?

**Reply by HUMORCE** [reply]
> the file won't be affected or removed if scoop happened to update to the latest version?

No, but it would be better to use alias/shim

`cp <Scoop>/apps/scoop/current/libexec/scoop-download.ps1 <Scoop>/shims/scoop-dl.ps1`, then modify paths:
```
. "$PSScriptRoot\..\lib\getopt.ps1"
. "$PSScriptRoot\..\lib\json.ps1" # 'autoupdate.ps1' (indirectly)
. "$PSScriptRoot\..\lib\autoupdate.ps1" # 'generate_user_manifest' (indirectly)
. "$PSScriptRoot\..\lib\manifest.ps1" # 'generate_user_manifest' 'Get-Manifest'
. "$PSScriptRoot\..\lib\install.ps1"
```

**Reply by Xelbayria** [reply]
> > the file won't be affected or removed if scoop happened to update to the latest version?
> 
> No, but it would be better to use alias/shim

ah, that's better opinion to use

> `cp <Scoop>/apps/scoop/current/libexec/scoop-download.ps1 <Scoop>/shims/scoop-dl.ps1`, then modify paths:
> 
> ```
> . "$PSScriptRoot\..\lib\getopt.ps1"
> . "$PSScriptRoot\..\lib\json.ps1" # 'autoupdate.ps1' (indirectly)
> . "$PSScriptRoot\..\lib\autoupdate.ps1" # 'generate_user_manifest' (indirectly)
> . "$PSScriptRoot\..\lib\manifest.ps1" # 'generate_user_manifest' 'Get-Manifest'
> . "$PSScriptRoot\..\lib\install.ps1"
> ```

modify paths with codeblock shown above?  i do not understand? 

# EDIT: 

nvm i figured out what you mean. and now it make no sense because I cannot find a way to change the cache_path in the `scoop-download.ps1`   

i'm confused because I thought the file is responsible for downloading the file to their location

**Reply by HUMORCE** [reply]
This just creates as the base for you to refactor it.

**Reply by Xelbayria** [reply]
> This just creates as the base for you to refactor it.

What do you mean? 

I wasn't able to find the variable "cache_path" to change the location where file will be saved to?

EDIT: 
ok, i figured it out. i don't realized that there was a bit more to the way system is set up. It's quite complicated and my head are killing me.. a bit of headaching. lol

I changed the parameter on calling function on line 105 via scoop-download file: `Invoke-CachedDownload $app $version $url -to "C:\Users\Username\Downloads" $manifest.cookie $false`

it worked great. just one solution to change what I need.   now I just need to make a few adjustment on the hash-checking.

**Reply by Xelbayria** [reply]
Just a quick question, 
do I need to change the folder to remove "Read Only"? I think that's the problem for the hash-check to fail because the file access is denied? 

And where is the code found so I don't need to include the URL when the file is downloaded?  like so: `filename#version#url.extension`   changed to `filename#version.extension`

**Reply by HUMORCE** [reply]
Follow what I said before, then you are using Scoop Core as a library.

This is an implementation that does not depend on Scoop Core:

dl.ps1:
```
$args | foreach {
  $manifest = scoop cat $_ | convertfrom-json
  if ($manifest.url) { $url = $manifest.url; $hash = $manifest.hash } else { $url = $manifest.'64bit'.url; $hash = $manifest.'64bit'.hash }
  $save_path = join-path "d:\downloads" (split-path -leaf $url)

  # start download and hash check.
  invoke-webrequest -uri $url -outfile $save_path
  ...
}
```

**Reply by Xelbayria** [reply]
> Follow what I said before, then you are using Scoop Core as a library.
> 
> This is an implementation that does not depend on Scoop Core:
> 
> dl.ps1:
> 

Ah. I understand now. Thanks. it looks like I'll need to look in how to implement the progress of downloading and hash-checking (OK or FAILED). 

EDIT: 
i've play around with `#/whatever`. I didn't realize that it is actually used to rename the downloaded file when it's moved to $dir. Because your code help me understand a few things that I do not think of.

**Comment by Xelbayria** [comment]
There is a package called "apkeep" in scoop which can be used to download the apk and output it to a different location. Pretty much similar to what I am looking. I've learned a lot of things from this discussion, thanks for helping out. 

I'll be closing this discussion.

**Comment by MysticSnows** [comment]
- `scoop help config` shows the following:
```
cache_path:
      For downloads, defaults to 'cache' folder under Scoop root directory.
```

- We can change this path using:
`scoop config cache_path <new_path>`
   - E.g.: `scoop config cache_path C:\Users\Username\Downloads` 
   *(Better to use double quotes if path contains blank space.)*

- **config.json** after this command:
   - `{ ... "cache_path": "C:\\Users\\Username\\Downloads" ... }`
   *(config.json is located in C:\Users\Username\.config\scoop)*
